pkg/insight/insightstore/deploymentstore.go

```diff
@@ -17,26 +17,362 @@ package insightstore
+	toYear := toTime.Year()
+
+	sub := toTime.Sub(fromTime)
+	if sub < 0 {
+		return nil, errInvalidArg
+	}
+	if sub > maxDuration {
+		return nil, errLargeDuration
+	}
+
+	var result []*model.InsightDeployment
+	for year := fromYear; year <= toYear; year++ {
+		dirPath := determineDeploymentDirPath(year, projectID)
+		meta := model.InsightDeploymentChunkMetadata{}
+		err := s.loadDataFromFilestore(ctx, fmt.Sprintf("%s/%s", dirPath, metaFileName), &meta)
+		if errors.Is(err, filestore.ErrNotFound) {
+			continue
+		}
+		if err != nil {
+			return nil, err
+		}
+
+		keys := meta.ListChunkNames(from, to)
+
+		for _, key := range keys {
+			data := model.InsightDeploymentChunk{}
+			err := s.loadDataFromFilestore(ctx, fmt.Sprintf("%s/%s", dirPath, key), &data)
+			if err != nil {
+				return nil, err
+			}
+
+			// Maybe we should use metadata instead of loading chunkdata
+			if data.Version >= minimumVersion {
+				deployments := data.ExtractDeployments(from, to)
+				result = append(result, deployments...)
+			}
+		}
+	}
+
+	return result, nil
+}
+
+// Put deployments must be sorted by CompletedAt,
+func (s *store) Put(ctx context.Context, projectID string, deployments []*model.InsightDeployment, version model.InsightDeploymentVersion) error {
+	dailyDeployments := insight.GroupDeploymentsByDaily(deployments, time.Local)
+
+	for _, daily := range dailyDeployments {
+		err := s.putDeployments(ctx, projectID, daily, version, time.Now())
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// deployments must be sorted by CompletedAt, and year must be same.
+// if deployments is too large, this function cannot store deployments efficiently.
+func (s *store) putDeployments(ctx context.Context, projectID string, deployments []*model.InsightDeployment, version model.InsightDeploymentVersion, updatedAt time.Time) error {
+	if len(deployments) == 0 {
+		return nil
+	}
+
+	var year = time.Unix(deployments[0].CompletedAt, 0).Year()
+
+	// Load chunk
+	dirPath := determineDeploymentDirPath(year, projectID)
+	meta := model.InsightDeploymentChunkMetadata{}
+
+	metaPath := fmt.Sprintf("%s/%s", dirPath, metaFileName)
+	err := s.loadDataFromFilestore(ctx, metaPath, &meta)
+	if err != nil && !errors.Is(err, filestore.ErrNotFound) {
+		return err
+	}
+
+	// create chunk and meta
+	if len(meta.Chunks) == 0 {
+		chunkKey := determineDeploymentChunkKey(1)
+		newMeta, newChunk, err := createNewChunkAndMeta(deployments, version, chunkKey, updatedAt)
+		if err != nil {
+			return err
+		}
+
+		_, err = s.saveDataIntoFilestore(ctx, metaPath, newMeta)
+		if err != nil {
+			return err
+		}
+
+		_, err = s.saveDataIntoFilestore(ctx, fmt.Sprintf("%s/%s", dirPath, chunkKey), newChunk)
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+
+	firstDeployment := deployments[0]
+
+	latestChunkMeta := meta.Chunks[len(meta.Chunks)-1]
+	if firstDeployment.CompletedAt < latestChunkMeta.To {
+		return errors.New("cannot overwrite past deployment")
```

Comment:
should we make this as defined error as others?

Reply:
I think we should define like others, I'll fix it